extensions/azurecore/src/account-provider/azureAccountProvider2.ts

```diff
@@ -0,0 +1,443 @@
+
+			this.handleAuthentication(code).catch(console.error);
+		});
+
+		pathMappings.set('/signin', initialSignIn);
+		pathMappings.set('/callback', callback);
+	}
+	private async makeWebRequest(accessToken: TokenResponse, uri: string): Promise<any> {
+		const params = {
+			headers: {
+				'Content-Type': 'application/json',
+				'Authorization': `Bearer ${accessToken.accessToken}`
+			},
+			json: true
+		};
+
+		return new Promise((resolve, reject) => {
+			request.get(uri, params, (error: any, response: request.Response, body: any) => {
+				const err = error ?? body.error;
+				if (err) {
+					return reject(err);
+				}
+				return resolve(body.value);
+			});
+		});
+	}
+
+	private async getTenants(userId: string, homeTenant: string): Promise<Tenant[]> {
+		const armToken = await this.getToken(userId, AzureAccountProvider.AadCommonTenant, this.metadata.settings.armResource.id);
+		const tenantUri = url.resolve(this.metadata.settings.armResource.endpoint, 'tenants?api-version=2015-01-01');
+		const armWebResponse: any[] = await this.makeWebRequest(armToken, tenantUri);
+
+		const promises = armWebResponse.map(async (value: { tenantId: string }) => {
+			const graphToken = await this.getToken(userId, value.tenantId, this.metadata.settings.graphResource.id);
+			let tenantDetailsUri = url.resolve(this.metadata.settings.graphResource.endpoint, value.tenantId + '/');
+			tenantDetailsUri = url.resolve(tenantDetailsUri, 'tenantDetails?api-version=2013-04-05');
+			const tenantDetails: any[] = await this.makeWebRequest(graphToken, tenantDetailsUri);
+
+			return {
+				id: value.tenantId,
+				userId: userId,
+				displayName: tenantDetailsUri.length && tenantDetails[0].displayName
+					? tenantDetails[0].displayName
+					: localize('azureWorkAccountDisplayName', "Work or school account")
+			} as Tenant;
+		});
+
+		const tenants = await Promise.all(promises);
+		const homeTenantIndex = tenants.findIndex(tenant => tenant.id === homeTenant);
+		if (homeTenantIndex >= 0) {
+			const homeTenant = tenants.splice(homeTenantIndex, 1);
+			tenants.unshift(homeTenant[0]);
+		}
+		return tenants;
+	}
+
+	/**
+	 * Authenticates an azure account and then emits an event
+	 * @param code Code from authenticating
+	 */
+	private async handleAuthentication(code: string): Promise<void> {
+		const token = await this.getTokenWithAuthCode(code, AzureAccountProvider.redirectUrlAAD);
+
+		const tenants = await this.getTenants(token.userId, token.userId);
+		let identityProvider = token.identityProvider;
+		if (identityProvider) {
+			identityProvider = identityProvider.toLowerCase();
+		}
+
+		// Determine if this is a microsoft account
+		let msa = identityProvider && (
+			identityProvider.indexOf('live.com') !== -1 || // lgtm [js/incomplete-url-substring-sanitization]
+			identityProvider.indexOf('live-int.com') !== -1 || // lgtm [js/incomplete-url-substring-sanitization]
+			identityProvider.indexOf('f8cdef31-a31e-4b4a-93e4-5f571e91255a') !== -1 ||
```

Comment:
Where are these values from?

Reply:
Just took them from azureAccountProivder.ts, I agree that they're a terrible way of checking for an account type.

If someone's domain is examplelive.com, it'd consider it a microsoft account. What's even weirder is that we're not taking into consideration that people can use whatever domain they want for their personal Microsoft account. Oh well, something that we should bring up with PM or other engineers.